persistence/src/main/java/de/cyface/persistence/PersistenceLayer.java

```diff
@@ -561,6 +565,96 @@ private void deletePoint3dData(final long measurementIdentifier) {
         }
     }
 
+    /**
+     * Returns the average speed of the measurement with the provided measurement identifier.
+     * <p>
+     * Loads the {@link Track}s from the database to calculate the metric on the fly [STAD-384].
+     *
+     * @param measurementIdentifier The id of the {@code Measurement} to load the track for.
+     * @param locationCleaningStrategy The {@link LocationCleaningStrategy} used to filter the
+     *            {@link GeoLocation}s
+     * @return The average speed in meters per second.
+     * @throws CursorIsNullException when accessing the {@code ContentProvider} failed
+     */
+    public double loadAverageSpeed(final long measurementIdentifier,
+            @NonNull final LocationCleaningStrategy locationCleaningStrategy) throws CursorIsNullException {
+
+        final List<Track> tracks = loadTracks(measurementIdentifier);
+        double speedSum = 0.0;
+        int speedCounter = 0;
+        for (final Track track : tracks) {
+            double sum = 0.0;
+            int counter = 0;
+            for (final GeoLocation location : track.getGeoLocations()) {
+                if (locationCleaningStrategy.isClean(location)) {
```

Comment:
I think it's expected, that the average speed does not decrease when the user is not moving for a short time. (also, because the customer said "should be like distance calculation)
Thus, I'm adding speed samples only after the location cleaning strategy accepts the location captured.

My understanding of the feature request was:

the average speed per measurement should be calculated
i.e. when we pause the first track of the measurement and resume, the speed samples of all sub-tracks are used to calculate the average speed
for this to work, we cannot just save "averageSpeed" but have to save the speed sum and counter.

Reply:
The customer did not say anything about "distance calculation". That is what I said. We should probably have a look into how other apps do this, before making a final decision.

Reply:
see discussion in Ticket

Reply:
Seems there is no easy answer to this. Lets got with this implementation then.